File: y15/d07/cpp/main.cpp
#include "helper.h"
#include <iostream>
#include <map>
#include <ostream>
#include <string>
#include <vector>
using namespace std;

int find_val(map<string, vector<string>> &instructions, const string &target) {
  map<string, int> values;
  regex pattern("\\d+");

  auto get_wire_val = [&pattern, &values](string wire) {
    auto m = re_match(wire, pattern);
    if (m.size() > 0) {
      return stoi(m[0]);
    }

    map<string, int>::iterator it = values.find(wire);
    if (it != values.end()) {
      return it->second;
    }

    return -1;
  };

  while (true) {
    bool resolved = true;
    for (map<string, vector<string>>::iterator it = instructions.begin();
         it != instructions.end(); it++) {
      string tid = it->first;
      vector<string> opr = it->second;
      string gate = opr[0], op1 = opr[1], op2 = opr[2];
      int v1 = get_wire_val(op1);
      if (v1 == -1) {
        resolved = false;
        continue;
      }

      if (gate == "") {
        values[tid] = v1;
      } else if (gate == "NOT") {
        values[tid] = 65535 - v1;
      } else {
        int v2 = get_wire_val(op2);
        if (v2 == -1) {
          resolved = false;
          continue;
        }

        if (gate == "AND") {
          values[tid] = v1 & v2;
        } else if (gate == "OR") {
          values[tid] = v1 | v2;
        } else if (gate == "LSHIFT") {
          values[tid] = v1 << v2 & 0xffff;
        } else if (gate == "RSHIFT") {
          values[tid] = v1 >> v2;
        }
      }
    }
    if (resolved) {
      break;
    }
  }

  return values[target];
}

int main(int argc, char *argv[]) {
  if (argc < 2) {
    cout << "Provide the input file name" << endl;
    return 1;
  }

  vector<string> lines = read_lines(argv[1]);
  map<string, vector<string>> instructions;
  for (const auto &line : lines) {
    auto splits = split_string(line, " -> ");
    string opr = splits[0], tid = splits[1];
    vector<string> temp = split_string(opr, " ");
    vector<string> ins;
    ins.push_back(temp.size() == 1 ? "" : temp.size() == 2 ? "NOT" : temp[1]);
    ins.push_back(temp.size() == 2 ? temp[1] : temp[0]);
    ins.push_back(temp.size() == 3 ? temp[2] : "");
    instructions[tid] = ins;
  }

  // part 1
  int part1 = find_val(instructions, "a");
  cout << "Part 1: " << part1 << endl;

  // part 2
  instructions["b"][1] = to_string(part1);
  int part2 = find_val(instructions, "a");
  cout << "Part 2: " << part2 << endl;

  return 0;
}

File: y15/d07/rust/src/main.rs
mod helper;

use helper::read_lines;
use std::{collections::HashMap, env};

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() > 1 {
        let file = &args[1];
        println!("File: {}", file);
        let lines = read_lines(file);
        solve(lines);
    } else {
        println!("No file name provided");
    }
}

fn get_wire_val(wire: &str, values: &HashMap<&str, i32>) -> i32 {
    match wire.parse::<i32>() {
        Ok(v) => v,
        Err(_) => match values.get(wire) {
            Some(v) => *v,
            None => -1,
        },
    }
}

fn solve(lines: Vec<String>) {
    let mut instructions: Vec<[String; 4]> = Vec::new();
    for line in lines {
        if let Some((opr, tid)) = line.split_once(" -> ") {
            let operation: Vec<&str> = opr.split(' ').collect();
            let ol = operation.len();
            let gate = match ol {
                2 => "NOT",
                3 => operation[1],
                _ => "",
            }
            .to_string();
            let op1 = match ol {
                2 => operation[1],
                _ => operation[0],
            }
            .to_string();
            let op2 = match ol {
                3 => operation[2],
                _ => "",
            }
            .to_string();

            instructions.push([tid.to_string(), gate, op1, op2])
        }
    }

    let part1 = get_value(&instructions, "a");
    println!("Part 1: {}", part1);

    for ins in &mut instructions {
        if ins[0] == "b" {
            ins[2] = part1.to_string();
            break;
        }
    }
    let part2 = get_value(&instructions, "a");
    println!("Part 2: {}", part2);
}

fn get_value(instructions: &Vec<[String; 4]>, target: &str) -> i32 {
    let mut values: HashMap<&str, i32> = HashMap::new();
    loop {
        let mut resolved = true;
        for [tid, gate, op1, op2] in instructions {
            let v1 = get_wire_val(op1, &values);
            if v1 == -1 {
                resolved = false;
                continue;
            }

            if gate.is_empty() {
                values.insert(tid, v1);
            } else if gate == "NOT" {
                values.insert(tid, 65535 - v1);
            } else {
                let v2 = get_wire_val(op2, &values);
                if v2 == -1 {
                    resolved = false;
                    continue;
                }
                match gate.as_str() {
                    "AND" => values.insert(tid, v1 & v2),
                    "OR" => values.insert(tid, v1 | v2),
                    "LSHIFT" => values.insert(tid, v1 << v2 & 0xffff),
                    _ => values.insert(tid, v1 >> v2),
                };
            };
        }

        if resolved {
            break;
        }
    }
    *values.get(target).unwrap_or(&-1)
}
